compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor.hs

```diff
@@ -413,6 +418,30 @@ checkKinds (GHC.L _ m) = do
         , "This will cause problems when importing this module via data-dependencies."
         ]
 
+-- | Set of package names and module name spaces to emit internal import warnings for
+-- e.g. ("daml-script", "Daml.Script") means that only the "daml-script" package can import "Daml.Script.*.Internal.*" without warning
+-- Module name spaces should be unique and non-overlapping
+warnInternalPackages :: [(LF.PackageName, GHC.ModuleName)]
+warnInternalPackages = fmap (bimap LF.PackageName GHC.mkModuleName)
+  [ ("daml3-script", "Daml.Script")
```

Comment:
We don't know the package name that the imports belong to at parse time, and at typechecking time, information about the original import statement is gone - distributed to all uses of it.
Instead, we work from module path "prefixes", which is at least good enough for Daml.Script. If we need something better in future, we can look into it, but it might need some ghc hacks